src/ocrd/processor/base.py

```diff
@@ -55,9 +55,36 @@ class ResourceNotFoundError(FileNotFoundError):
     def __init__(self, name, executable):
         self.name = name
         self.executable = executable
-        self.message = "Could not find resource '%s' for executable '%s'. " \
-                       "Try 'ocrd resmgr download %s %s' to download this resource." \
-                       % (name, executable, executable, name)
+        self.message = (f"Could not find resource '{name}' for executable '{executable}'. "
+                        f"Try 'ocrd resmgr download {executable} {name}' to download this resource.")
+        super().__init__(self.message)
+
+class NonUniqueInputFile(ValueError):
+    """
+    An exception signifying the specified fileGrp / pageId / mimetype
+    selector yields multiple PAGE files, or no PAGE files but multiple images,
+    or multiple files of that mimetype.
+    """
+    def __init__(self, fileGrp, pageId, mimetype):
+        self.fileGrp = fileGrp
+        self.pageId = pageId
+        self.mimetype = mimetype
+        self.message = (f"Could not determine unique input file for fileGrp {fileGrp} "
+                        f"and pageId {pageId} under mimetype {mimetype or 'PAGE+image(s)'}")
+        super().__init__(self.message)
+
+class MissingInputFile(ValueError):
+    """
+    An exception signifying the specified fileGrp / pageId / mimetype
+    selector yields no PAGE files, or no PAGE and no image files,
+    or no files of that mimetype.
+    """
+    def __init__(self, fileGrp, pageId, mimetype):
+        self.fileGrp = fileGrp
+        self.pageId = pageId
+        self.mimetype = mimetype
+        self.message = (f"Could not find input file for fileGrp {fileGrp} "
+                        f"and pageId {pageId} under mimetype {mimetype or 'PAGE+image(s)'}")
```

Comment:
Appreciate the specific exceptions. Maybe we should keep all of those in a separate `ocrd_utils.exceptions` before we release 3.0 stable.

Reply:
Is there a case where `ocrd` module code will be used inside `ocrd_utils.exceptions` and then that code is imported back to `ocrd` from `ocrd_utils`? IMO, no circularity should happen if implemented properly, i.e., just importing the required definitions or methods of a class instead of `*`. This said dot operations in Python are also more expensive in terms of performance.

```python
import workspace
workspace.add_file()
```

is less efficient than

```python
from workspace import add_file
add_file()
```

due to the extra attribute lookup overhead (usually bad inside loops).

Reply:
Interesting about the import syntax. I prefer the latter anyway, good to hear that it's also faster.

> Wouldn't that defeat package encapsulation (and possibly lead to circularity)? If I use stuff from ocrd which throws an exception, which I want to be able to handle, then that should be in ocrd as well.

The dependency graph is `ocrd_utils -> ocrd_models -> ocrd_validators -> ocrd -> ocrd_network`. So unless `ocrd_utils.exceptions` imports something from one of the other packages (like the problem with in https://github.com/bertsky/core/pull/12), there won't be an issue, at least for these simple, builtin-derived exceptions. If we want e.g. `JsonSchemaDeprecationWarning` in there as well, then we'd have to move it up the package hierarchy, e.g. to `ocrd.exceptions`. Which would not work for `JsonSchemaDeprecationWarning` because that is used in `ocrd_validators`.

It is not important, we just have quite a few custom exceptions and I thought it would be sensible to have them in one place, as many python projects do. But let's put a pin into that discussion until we're ready to release, there's more pressing issues.